viewer/cache.cpp

```diff
@@ -0,0 +1,74 @@
+#include "cache.hpp"
+
+#include <hawktracer/base_types.h>
+
+namespace HawkTracer
+{
+namespace viewer
+{
+
+Cache::Cache()
+{
+}
+
+bool Cache::range_in_cache(HT_TimestampNs start_ts,
+                           HT_TimestampNs stop_ts,
+                           HT_EventKlassId klass_id)
+{
+    if (_events.find(klass_id) == _events.end())
+    {
+        return false;
+    }
+    HT_TimestampNs cache_start_ts = _events[klass_id].front().get().get_timestamp();
+    HT_TimestampNs cache_stop_ts = _events[klass_id].back().get().get_timestamp();
+    if (cache_start_ts != start_ts || cache_stop_ts != stop_ts)
+    {
+        return false;
+    }
+    return true;
+}
+
+std::vector<EventRef> Cache::get_data(HT_EventKlassId klass_id)
+{
+    return _events[klass_id];
```

Comment:
If you search for an unknown klass_id here using operator[] it will create an empty vector in _events (hence method is non const).

The at() or find() methods on unordered_map lets you do const lookups. I'd consider using one of those and handling unknown keys somehow, not sure on rules with exceptions on this lib but could return a std::optional

Reply:
I couldn't use std::optional and I think it's because C++11 is used for building. But, since we already link the boost library, I've changed this method to return boost::optional.